// @parcel/reporter-dev-server/lib/ServerReporter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _plugin = require("@parcel/plugin");

var _Server = _interopRequireDefault(require("./Server"));

var _HMRServer = _interopRequireDefault(require("./HMRServer"));

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

let servers = new Map();
let hmrServers = new Map();

var _default = new _plugin.Reporter({
  async report({
    event,
    options,
    logger
  }) {
    let serve = options.serve;
    if (!serve) return;
    let server = servers.get(serve.port);
    let hmrServer = hmrServers.get(serve.port);

    switch (event.type) {
      case 'watchStart':
        {
          var _serve$publicUrl;

          // If there's already a server when watching has just started, something
          // is wrong.
          if (server) {
            return logger.warn({
              message: 'Trying to create the devserver but it already exists.'
            });
          }

          let serverOptions = _objectSpread({}, serve, {
            projectRoot: options.projectRoot,
            cacheDir: options.cacheDir,
            distDir: _path.default.join(options.cacheDir, 'dist'),
            // Override the target's publicUrl as that is likely meant for production.
            // This could be configurable in the future.
            publicUrl: (_serve$publicUrl = serve.publicUrl) !== null && _serve$publicUrl !== void 0 ? _serve$publicUrl : '/',
            inputFS: options.inputFS,
            outputFS: options.outputFS,
            logger
          });

          server = new _Server.default(serverOptions);
          servers.set(serve.port, server);
          const devServer = await server.start();

          if (options.hot) {
            const hmrServer = new _HMRServer.default({
              devServer,
              logger
            });
            hmrServers.set(serve.port, hmrServer);
            hmrServer.start();
          }

          break;
        }

      case 'watchEnd':
        if (!server) {
          return logger.warn({
            message: 'Could not shutdown devserver because it does not exist.'
          });
        }

        if (hmrServer) {
          hmrServer.stop();
        }

        await server.stop();
        servers.delete(serve.port);
        hmrServers.delete(serve.port);
        break;

      case 'buildSuccess':
        if (!server) {
          return logger.warn({
            message: 'Could not send success event to devserver because it does not exist.'
          });
        }

        server.buildSuccess(event.bundleGraph);

        if (hmrServer) {
          hmrServer.emitUpdate(event);
        }

        break;

      case 'buildFailure':
        // On buildFailure watchStart sometimes has not been called yet
        // do not throw an additional warning here
        if (!server) return;
        server.buildError(event.diagnostics);

        if (hmrServer) {
          hmrServer.emitError(event.diagnostics);
        }

        break;
    }
  }

});

exports.default = _default;
